File: sdk/python/provenance_client/core/session.py
import logging
from typing import Any, Callable, TypeVar

from ..interfaces import ProvenanceGatewayProtocol, ProvenanceSessionProtocol
from .models import Decision, ExecutionResult

F = TypeVar("F", bound=Callable[..., Any])
log = logging.getLogger(f"sdk.{__name__}")


class ProvenanceSession(ProvenanceSessionProtocol):
    """
    Scopes a series of tool-call executions to a shared session_id.

    Supports use as a sync context manager.  For async, use async with directly on the ProvenanceClient and pass session_id= explicitly.
    """

    def __init__(
        self,
        gateway: ProvenanceGatewayProtocol,
        session_id: str,
    ) -> None:
        self._gateway = gateway
        self.session_id = session_id
        self._results: list[ExecutionResult] = []

    # ── execute variants ──────────────────────────────────────────────────────

    def execute(
        self,
        action: str,
        parameters: dict[str, Any],
        *,
        decision: Decision | None = None,
    ) -> ExecutionResult:
        result = self._gateway.execute(
            action,
            parameters,
            session_id=self.session_id,
            decision=decision,  # type: ignore
        )
        self._results.append(result)  # type: ignore
        return result

    async def async_execute(
        self,
        action: str,
        parameters: dict[str, Any],
        *,
        decision: Decision | None = None,
    ) -> ExecutionResult:
        result = await self._gateway.async_execute(
            action,
            parameters,
            session_id=self.session_id,
            decision=decision,
        )
        self._results.append(result)  # type: ignore
        return result

    # ── guard scoped to this session ──────────────────────────────────────────

    def guard(self, action: str | None = None, **kwargs) -> Callable[[F], F]:
        """Same as ProvenanceClient.guard() but pins session_id to this session."""
        return self._gateway.guard(action, session_id=self.session_id, **kwargs)

    # ── session summary ───────────────────────────────────────────────────────

    @property
    def results(self) -> list[ExecutionResult]:
        return list(self._results)

    @property
    def blocked_count(self) -> int:
        return sum(1 for r in self._results if r.blocked)

    @property
    def allowed_count(self) -> int:
        return sum(1 for r in self._results if r.allowed)

    # ── context manager ───────────────────────────────────────────────────────

    def __enter__(self) -> "ProvenanceSession":
        return self

    def __exit__(self, *_) -> None:
        log.info(
            f"Session {self.session_id!r} completed with {len(self._results)} results"
        )

    def __repr__(self) -> str:
        return (
            f"ProvenanceSession(id={self.session_id!r}, results={len(self._results)})"
        )

File: sdk/typescript/provenance_client/core/session.ts
import { Decision, ExecutionResult } from "./models";
import { ProvenanceGateway } from "./gateway";
import { ProvenanceSessionProtocol } from "../interfaces/session";

export class ProvenanceSession implements ProvenanceSessionProtocol {
  public readonly sessionId: string;
  private readonly gateway: ProvenanceGateway;
  private readonly _results: ExecutionResult[] = [];

  constructor(gateway: ProvenanceGateway, sessionId: string) {
    this.gateway = gateway;
    this.sessionId = sessionId;
  }

  public execute(
    action: string,
    parameters: Record<string, any>,
    options: {
      decision?: Decision;
    } = {},
  ): ExecutionResult {
    throw new Error(
      "Synchronous execution is not supported for ProvenanceSession. Use asyncExecute() instead.",
    );
  }

  public async asyncExecute(
    action: string,
    parameters: Record<string, any>,
    options: {
      decision?: Decision;
    } = {},
  ): Promise<ExecutionResult> {
    const result = await this.gateway.asyncExecute(action, parameters, {
      sessionId: this.sessionId,
      decision: options.decision,
    });
    this._results.push(result);
    return result;
  }

  public guard(
    action?: string,
    options: {
      decision?: Decision;
      raiseOnBlock?: boolean;
    } = {},
  ): <T extends (...args: any[]) => any>(func: T) => T {
    return this.gateway.guard(action, {
      ...options,
      sessionId: this.sessionId,
    });
  }

  public get results(): ExecutionResult[] {
    return [...this._results];
  }

  public get blockedCount(): number {
    return this._results.filter((r) => r.blocked).length;
  }

  public get allowedCount(): number {
    return this._results.filter((r) => r.allowed).length;
  }

  public [Symbol.dispose](): void {
    console.log(
      `Session ${this.sessionId} completed with ${this._results.length} results`,
    );
  }

  public toString(): string {
    return `ProvenanceSession(id=${this.sessionId}, results=${this._results.length})`;
  }
}
